### ISSUE TITLE
Problems about the implementation

### ISSUE THREAD

**Issue by Junction4Nako** [eric-ai-lab/MiniGPT-5#29]
Thank you very much for sharing such fascinating work. I've encountered some questions regarding implementation details while going through the code. I hope to get your insights.
1. during the training procedure, the generation targets of an image are "[IMG0][IMG1][IMG3]...[IMG7]"? However, I found that the model will only generate [IMG0] during the evaluation procedure, did I make any mistake?
2. I have another question about the "hidden states output" of the generative model, in the Line 307-135 of model.py
`        
        predicted_images_ft = None
        if len(special_token_index):
            idx = special_token_index[0,0]
            t2i_input_embedding = last_hidden_state[idx][-1]
            assert t2i_input_embedding.shape[1] == self.img_token_num  # HERE for the reference
            img0_output_feature = last_hidden_state[idx-1][-1][:, -1:]
            t2i_input_embedding = torch.cat([img0_output_feature, t2i_input_embedding[:, :-1]], dim=1)
            t2i_input_embedding = self.fc(t2i_input_embedding)
            mapping_feature = self.llm_to_t2i_mapping(src=t2i_input_embedding, tgt=self.t2i_decoder_prompt)
`
I wonder why last_hidden_state[idx][-1] (the position of [IMG0]) always has a shape of [1, 8, 4096]?

**Comment by KzZheng** [comment]
Thanks for your attention! To answer your question
1. Only show [IMG0] in the text output is correct. Currently, we will automatically generate "[IMG1][IMG3]...[IMG7]" when we detect the [IMG0] is generated.
2. Same answer as below; we will automatically generate the remaining vokens. You can find related code under this condition [Line 324](https://github.com/eric-ai-lab/MiniGPT-5/blob/cdd6f6f90b3dfd5dab312b92b25e4a7bb5db8843/minigpt4/models/mini_gpt5.py#L324).

**Comment by prefixRAINSTARsuffix** [comment]
> Thanks for your attention! To answer your question
> 
> 1. Only show [IMG0] in the text output is correct. Currently, we will automatically generate "[IMG1][IMG3]...[IMG7]" when we detect the [IMG0] is generated and we will not show all of them in the output text.
> 2. Same answer as above; we will automatically generate the remaining vokens. You can find related code under this condition [Line 324](https://github.com/eric-ai-lab/MiniGPT-5/blob/cdd6f6f90b3dfd5dab312b92b25e4a7bb5db8843/minigpt4/models/mini_gpt5.py#L324).

@KzZheng 请问这样的话，是不是每次推理，输入SD的feature，只有 [IMG0] 的 feature 是可变的，[IMG1][IMG2]...[IMG7] 的 feature永远是固定的

**Comment by prefixRAINSTARsuffix** [comment]
> Thanks for your attention! To answer your question
> 
> 1. Only show [IMG0] in the text output is correct. Currently, we will automatically generate "[IMG1][IMG3]...[IMG7]" when we detect the [IMG0] is generated and we will not show all of them in the output text.
> 2. Same answer as above; we will automatically generate the remaining vokens. You can find related code under this condition [Line 324](https://github.com/eric-ai-lab/MiniGPT-5/blob/cdd6f6f90b3dfd5dab312b92b25e4a7bb5db8843/minigpt4/models/mini_gpt5.py#L324).

@KzZheng  In this case, only the output feature of [IMG0] is changeable during each inference, and the output feature of [IMG1][IMG2]...[IMG7] is fixed. So among the 8 features that prompting SD, only the first one is changeable. I'm wondering if I understand correctly.

**Comment by KzZheng** [comment]
No. All output features are changeable during inference. For the original LLM, the model always takes the previous output token as the new input token. Here, you can imagine that we only fix the input tokens with [IMG1]...[IMG7], when we detect the previous output token is [IMG0]. But the output features always depend on all previous tokens.
